### PR TITLE
sql: run deleteRange queries in batches

### LINKED ISSUE #36588: memory usage spike with range delete
**Describe the problem**

When I do `DELETE FROM db.foo` on a large table, I see memory usage of cockroach spike.

**To Reproduce**

8 node cluster. Table db.foo has 1 million keys (each key is an int), total disk usage 26 GiB. 
When I tried `DELETE FROM db.foo`, memory usage on the gateway node went up by 2 GB.

Is that command loading all keys in memory before issuing deletes? The command was fast enough that I believe that it does not put an individual tombstone on each row, but the memory usage is unexpected.

TRUNCATE does not seem to exhibit this problem. I could be using TRUNCATE here, but TRUNCATE is treated like a schema change for privilege purposes which is giving us some trouble given the way we setup privileges in our services.

**Environment:**
Cockroach version 2.1.3
Server Linux
Client `cockroach sql`

### PR DESCRIPTION
**Pull request by jordanlewis**
Fixes #36588.

Previously, a query like `DELETE FROM t` that uses a deleteRangeNode
under the hood would attempt to send a single `DelRange` request for
each span it was responsible for deleting. This led to unlimited memory
usage, since the `DelRange` request must return the keys it deleted in
this context so SQL can know how many rows were affected.

To solve this problem, we send `DelRange` requests with a returned-key
limit in a loop until the range is fully deleted.

Release note (bug fix): prevent unlimited memory usage during SQL range
deletions

### CODE REVIEW

**Comment by jordanlewis** [comment]
I deleted a test that was testing whether or not `DELETE FROM t` produced a parallelized batch. It no longer does. We already test for the presence of parallelized batches from the distsender in another trace test, so deleting this test doesn't reduce coverage.

**Review by BramGruneir — DISMISSED** [review]
So the technique you used here isn't used anywhere else in our code (that I could find).  This is a different type of batching than what we use for other writers.

Since this is a fast path, and it's already specialized, would it make sense to using a range_delete instead of point deletes if the size is large enough to batch?  

This would avoid the batching and most likely be more efficient.  

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-:-Lcfjb2b61avpSjJPagq:b-hohsz0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @jordanlewis)

___
*[pkg/sql/tablewriter.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcfiGLs4YNQPydissyb:-LcfiGLs4YNQPydissyc:bcvy1og) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/tablewriter.go#L95)):*
> ```Go
> )
> 
> // extendedTableWriter is a temporary interface introduced
> ```

This old TODO (that is missing a todo) should have a github issue associated with it. 
___
*[pkg/sql/logictest/testdata/planner_test/delete_range, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-Lcfhk4q2WC_Tc2V-KW4:-Lcfhk4q2WC_Tc2V-KW5:baosc45) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/logictest/testdata/planner_test/delete_range#L24)):*
> ```
>   AND operation != 'dist sender send'
> ----
> flow                 DelRange /Table/53/1 - /Table/53/2
> ```

Won't this test be brittle to new system tables being added?  
We really need a way  to put in wildcards in the results.

**Review by andreimatei — CHANGES_REQUESTED** [review]
Bram I probably don't understand you comment, but just in case - this patch batches less than before, not more - by putting a limit and retrying. And it already uses range deletes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-:-LcrSlyLDDbTX03EYv7o:b9nxhgk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir and @jordanlewis)

___
*[pkg/sql/delete_range.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrMVKpD1Jpiz1ovxEQ:-LcrMVKpD1Jpiz1ovxER:bxt7cee) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L171)):*
> ```Go
> 	traceKV := params.p.ExtendedEvalContext().Tracing.KVTracingEnabled()
> 	for len(d.spans) != 0 {
> 		d.tableWriter.init(params.p.txn)
> ```

why are we using a tablewriter here? Seems that we're dealing with key ranges, not rows. What's the tablewriter giving us? Let's control our own destiny if possible.
___
*[pkg/sql/delete_range.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrP8vM0DFexI7KYa1a:-LcrP8vM0DFexI7KYa1b:b-vqn7hw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L184)):*
> ```Go
> 		}
> 
> 		d.spans = d.spans[:0]
> ```

comment that we're going to add all the spans we haven't touched back. And also I'd operate on a copy of d.spans; there's no reason for d.spans to be updated.

But better yet, I think the code would be clearer if you iterated over the results (which should be in order) and remove from the head of d.spans items one by one while there's no `ResumeSpan` (instead of wiping d.spans and then reconstructing it)
___
*[pkg/sql/delete_range.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrNyl--Lnzr0UahKMT:-LcrNyl09puCFQgqUTqe:b7smxyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L187)):*
> ```Go
> 		for _, r := range d.tableWriter.b.Results {
> 			var prev []byte
> 			for _, i := range r.Keys {
> ```

the name `i` here seems particularly bad. Mind changing it?
___
*[pkg/sql/tablewriter.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrLW2u1FvAxfvVI0yn:-LcrLW2vAkc75Wwdhoka:b-hkr3ri) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/tablewriter.go#L129)):*
> ```Go
> 	txn *client.Txn
> 	// is autoCommit turned on.
> 	autoCommit autoCommitOpt
> ```

this is lolo. Make it a bool :)
___
*[pkg/sql/logictest/testdata/planner_test/delete_range, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrQ1wF1kwfAkfvFyjU:-LcrQ1wF1kwfAkfvFyjV:b-4ygdk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/logictest/testdata/planner_test/delete_range#L13)):*
> ```
> SET tracing = on,kv; DELETE FROM a; SET tracing = off
> 
> # Ensure that DelRange requests are chunked for DELETE FROM...
> ```

If all you care about are the `DelRange` messages, delete all that filtering crap and say `message like 'DelRange%'`.
If you care about more than those messages, I for one can't support writing more tests like this - versus writing the tests in go and using a more proper library for asserting whatever.
___
*[pkg/sql/logictest/testdata/planner_test/delete_range, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-Lcfhk4q2WC_Tc2V-KW4:-LcrQD-w94dMt3I_T6yg:b-a91dv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/logictest/testdata/planner_test/delete_range#L24)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Won't this test be brittle to new system tables being added?  
We really need a way  to put in wildcards in the results.
</blockquote></details>

there's room for new system tables below id 50.

**Review by jordanlewis — COMMENTED** [review]
Bram, I also don't really understand your comment. The purpose of this patch is to break up a single enormous range delete into many smaller ones by imposing a limit on the number of keys that the range delete can delete. This batching method is used elsewhere in the codebase, namely in scans - we might ask for a huge span of keys, but we set a limit on the number of keys returned from the scan to prevent the KV layer from allocating unlimited memory and killing itself.

If you're suggesting we use a *clear* range, we can't, because this has to be transactional and respect MVCC.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-:-LcvBibi31M5D1OyvFYz:b2xr5ge)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @BramGruneir)

___
*[pkg/sql/delete_range.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrMVKpD1Jpiz1ovxEQ:-LcvCNUFBxF4hjTP-Ddq:bkytajc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L171)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why are we using a tablewriter here? Seems that we're dealing with key ranges, not rows. What's the tablewriter giving us? Let's control our own destiny if possible.
</blockquote></details>

It's convenient, provides error handling, and is the same as what we do everywhere else. I don't see a good reason to make this code less DRY.
___
*[pkg/sql/delete_range.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrP8vM0DFexI7KYa1a:-LcvGWzQ3JqiT4vsToMW:b5jmq8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L184)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment that we're going to add all the spans we haven't touched back. And also I'd operate on a copy of d.spans; there's no reason for d.spans to be updated.

But better yet, I think the code would be clearer if you iterated over the results (which should be in order) and remove from the head of d.spans items one by one while there's no `ResumeSpan` (instead of wiping d.spans and then reconstructing it)
</blockquote></details>

There's still the issue that you have to write the resume span back to the spans array. I've tried something like your approach, what do you think?
___
*[pkg/sql/delete_range.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrNyl--Lnzr0UahKMT:-LcvEhMI3xzKcgwlRr5p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L187)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the name `i` here seems particularly bad. Mind changing it?
</blockquote></details>

Done.
___
*[pkg/sql/tablewriter.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcfiGLs4YNQPydissyb:-LcvExIu1GMKDf7GbqJI:b-j5mf5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/tablewriter.go#L95)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This old TODO (that is missing a todo) should have a github issue associated with it. 
</blockquote></details>

Really? Most TODOs in our codebase don't have github issues associated with it, not sure why this one would be special
___
*[pkg/sql/tablewriter.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrLW2u1FvAxfvVI0yn:-LcvEjJL91PmTtESbpdD:b2zf3v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/tablewriter.go#L129)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this is lolo. Make it a bool :)
</blockquote></details>

I'd rather not do that in this commit.
___
*[pkg/sql/logictest/testdata/planner_test/delete_range, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrQ1wF1kwfAkfvFyjU:-LcvEszt80Y3s0KwZUWG:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/logictest/testdata/planner_test/delete_range#L13)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If all you care about are the `DelRange` messages, delete all that filtering crap and say `message like 'DelRange%'`.
If you care about more than those messages, I for one can't support writing more tests like this - versus writing the tests in go and using a more proper library for asserting whatever.

</blockquote></details>

Good point, done.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @BramGruneir)

___
*[pkg/sql/delete_range.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrP8vM0DFexI7KYa1a:-LcvKRnL2u1X-s6Oi1ct:bt42ehl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L184)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

There's still the issue that you have to write the resume span back to the spans array. I've tried something like your approach, what do you think?
</blockquote></details>

Eh, the approach that I took is subtly buggy. I think the way that I had originally is simpler - going to revert to it.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @BramGruneir)

___
*[pkg/sql/delete_range.go, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrP8vM0DFexI7KYa1a:-LcvKnMr9ElR7YFSY20d:b-c1s1b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L184)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Eh, the approach that I took is subtly buggy. I think the way that I had originally is simpler - going to revert to it.
</blockquote></details>

Kept your suggestion to copy the spans though.

**Review by andreimatei — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @BramGruneir, and @jordanlewis)

___
*[pkg/sql/delete_range.go, line 171 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LcrMVKpD1Jpiz1ovxEQ:-LcvPtdkDAwa7HrkKXhV:brigqkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2385c31a1ff4412719ecc6058915fbf649755deb/pkg/sql/delete_range.go#L171)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's convenient, provides error handling, and is the same as what we do everywhere else. I don't see a good reason to make this code less DRY.
</blockquote></details>

what convenience? You're calling `init` repeatedly, `finalize` (at least can you call `flushAndStartNewBatch`?) repeatedly, you're manually reaching into `d.tableWriter.b.Header`, you're reaching into `autoCommit`.
What error handling? `row.ConvertBatchError` deals with uniqueness constraint violations, which are not applicable here. Also note how that function is in the `row` package and we're not operating on rows here.
Tablewriters, as I understand them, have something to do with rows, and we don't seem to be using that here. I'm not completely sure why not, btw - we seem to be doing our own counting of rows; can't the tw do that? Cause if we used it for that, then I'd be for it.

So basically as far as I can see, what you want here is simply to use the KV API. Using the KV API directly (which means `txn.Batch()`, `txn.Run()`) is not "repeating yourself".

**Comment by jordanlewis** [comment]
@andreimatei RFAL, I've removed the `tableWriter` - you were right.

**Review by andreimatei — APPROVED** [review]
:LGTM:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-:-LfWXOrDAvJ0nbv2GXlm:bmxilld)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @BramGruneir, and @jordanlewis)

___
*[pkg/sql/delete_range.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36728#-LfWScgOAldJExIuKg3I:-LfWScgOAldJExIuKg3J:b-an4hyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c668645181c145519f29f6b7b9617e5299d03b8/pkg/sql/delete_range.go#L231)):*
> ```Go
> func (d *deleteRangeNode) enableAutoCommit() {
> 	// DeleteRange can't autocommit, because it has to delete in batches, and it
> 	// won't know whether or not there is more work to do until after the first
> ```

nit: it's not about the first range in particular, is it?

**Comment by jordanlewis** [comment]
TFTR!

bors r+
